src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,147 @@
+            size++;
+            return;
+        }
+        if (index == 0) {
+            Node<T> nodeSecond = head;
+            node = new Node<>(null,value, nodeSecond);
+            nodeSecond.prev = node;
+            head = node;
+            size++;
+            return;
+        }
+        Node<T> firstNode = nodeForIndex(index - 1);
+        Node<T> secondNode = null;
+        if (size > index) {
+            secondNode = nodeForIndex(index);
+        }
+        Node<T> addedNode = new Node<>(firstNode, value, secondNode);
+        firstNode.next = addedNode;
+        if (size == index) {
+            tail = addedNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T nodeValue : list) {
+            add(nodeValue);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return nodeForIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> firstNode = nodeForIndex(index);
+        T oldValue = firstNode.value;
+        firstNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> removedNode = nodeForIndex(index);
+        size--;
+        return unlink(removedNode);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> result = head;
+        for (int i = 0; i < size;i++) {
+            if (result.value == object || result.value.equals(object)) {
+                size--;
+                unlink(result);
+                return true;
+            }
+            result = result.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkPositionIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private Node<T> nodeForIndex(int index) {
+        checkPositionIndex(index);
+        Node<T> newNode = head;
+        for (int i = 0; i < index; i++) {
+            newNode = newNode.next;
+        }
+        return newNode;
+    }
+
+    private T unlink(Node<T> node) {
+        Node<T> firstNode = node.prev;
+        Node<T> secondNode = node.next;
+        node = new Node<>(null, node.value, null);
+        if (firstNode == null && secondNode == null) {
+            return node.value;
+        }
+
+        if (firstNode == null) {
+            secondNode.prev = firstNode;
+            head = secondNode;
+            return node.value;
+        }
+        if (secondNode == null) {
+            firstNode.next = secondNode;
+            return node.value;
+        }
+        firstNode.next = secondNode;
+        secondNode.prev = firstNode;
+        return node.value;
```

Comment:
```suggestion
        if (firstNode == null) {
            head = secondNode;
        } else {
            firstNode.next = secondNode;
            node.prev = null;
        }
        if (secondNode == null) {
            tail = firstNode;
        } else {
            secondNode.prev = firstNode;
            node.next = null;
        }
        node.value = null;
        return element;
```
we could simplify this